### ISSUE TITLE
[clang-tidy] Clang-Tidy produces a false positive when returning from a lambda `cppcoreguidelines-owning-memory`

### ISSUE THREAD

**Issue by H-G-Hristov** [llvm/llvm-project#59389]
Clang-Tidy produces a false positive when returning from a lambda:

https://releases.llvm.org/15.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/cppcoreguidelines/owning-memory.html#cppcoreguidelines-owning-memory

> returning a newly created resource of type 'void' or 'gsl::owner<>' from a function whose return type is not 'gsl::owner<>'C/C++[cppcoreguidelines-owning-memory](https://releases.llvm.org/15.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/cppcoreguidelines/owning-memory.html)

Is caused by:

```C++
TEST(ExternalLibTest, MicrosoftGslOwner)
{
    struct S
    {
        int value{1};
    };

    const auto MakeS = [] -> ::gsl::owner<S*>
    {
        return ::gsl::owner<S*>{new S{}};
    };

    const ::gsl::owner<S*> sPtr = MakeS();

    EXPECT_EQ(sPtr->value, 1);
    EXPECT_NE(sPtr->value, 2);
}
```

but is fine:

```C++
namespace {

struct S
{
    int value{1};
};

auto MakeInt()
{
    return ::gsl::owner<S*>{new S{}};
}

}  // namespace
```

**Comment by PiotrZSL** [comment]
https://reviews.llvm.org/D157285
